### PR TITLE
 [PASS] Improve storage rewrite, support reuse for different dtypes 

### PR DESCRIPTION
**Pull request by xqdan**

### CODE REVIEW

File: src/pass/storage_rewrite.cc [hunk 1]
```diff
@@ -491,7 +491,7 @@ class StoragePlanRewriter : public IRMutator {
   }
   // Remap the index
   Expr RemapIndex(Type dtype, Expr index, StorageEntry* e) {
-    CHECK_EQ(dtype.element_of(), e->elem_type);
+    CHECK_LE(dtype.element_of().bytes(), e->elem_type.bytes());
     if (e->elem_offset == 0) return index;
     return make_const(index.type(), e->elem_offset) + index;
```
Comment (tqchen) [inline src/pass/storage_rewrite.cc]: if bytes are different, the index calculation rule would need to change.

**Comment by tqchen** [comment]
The current rule does not seem to be correct for elem_offset calculation for different data types, because the index calculation need to be translated among different types(they are in unit of access element type)

**Comment by tqchen** [comment]
cc @libing4752  who looked at this issue before

File: src/pass/storage_rewrite.cc (hunk 1 above)
Comment (xqdan) [reply src/pass/storage_rewrite.cc]: say we have
a[] = 1.0 (float32)
b[] = 1  (int16)
c[] = 1  (int 16)
d[] = b[] + c[]

do you mean?
b -> a[0] elem_offset = 0
c -> a[16] elem_offset = 16

**Comment by tqchen** [comment]
If we have 
```c++
int32 a[10];
int64 b[10];

b[i] = ...

```
after merging b after a, it should become (note cast and index is semantics of access ptr)
```c++
int32 a_b_combo[10 + 20];

(cast<int64*>(a_b_combo))[5 + i] = ...
```

**Comment by xqdan** [comment]
should we unify all types or just unify types for space reusing vars?

**Comment by tqchen** [comment]
didn't get what you mean, maybe provide an example proposal? The current definition of access_ptr is fine as long as we do it right and respect its definition

**Comment by xqdan** [comment]
i mean for case below, we may not need to unify int64 as int32
```
int32 a[10] = 1;
int64 b[10] = 2;
int32 c[10] = a[10];
```

for this case, we can unify int64 as int32
```
int32 a[10] = 1;
int64 b[5] = 2;
```
for this case, b[10] is larger than a[10], it would alloc another space for b[10] for current logic
```
int32 a[10] = 1;
int64 b[10] = 2;
```

**Comment by kun-zh** [comment]
Use a_b_combo.access_ptr(ptr_type="int64")? But for my understanding, it should generate "cast<int64>" not "cast<int64*>".

**Comment by xqdan** [comment]
that's pyton's api. what i need to do is make that ir generate in ir mutate, in which i don't know how to do that

**Comment by xqdan** [comment]
for this case:
```
def test_alloc_seq_type():
    ib = tvm.ir_builder.create()
    n = tvm.var("n")
    with ib.for_range(0, n, name="i") as i:
        with ib.for_range(0, 10, name="j") as j:
            A = ib.allocate("float32", 200, name="A", scope="local.L0A")
            A[j] = 1.2
        with ib.for_range(0, 10, name="j") as j:
            B = ib.allocate("int16", 200, name="B", scope="local.L0A")
            B[j] = tvm.const(1, "int16")
        with ib.for_range(0, 10, name="j") as j:
            C = ib.allocate("int16", 200, name="C", scope="local.L0A")
            C[j] = tvm.const(1, "int16")
        with ib.for_range(0, 10, name="j") as j:
            D = ib.allocate("int16", 200, name="D", scope="local.L0A")
            D[j] = C[j] + B[j]

    body = ib.get()
    body = tvm.ir_pass.StorageRewrite(body)
```
we can get ir like 
```
// attr [A] storage_scope = "local.L0A"
allocate A[float32 * 300]
for (i, 0, n) {
  for (j, 0, 10) {
    A[j] = 1.200000f
  }
  for (j, 0, 10) {
    A[j] = (int16)1
  }
  for (j, 0, 10) {
    A[(400 + j)] = (int16)1
  }
  for (j, 0, 10) {
    A[(400 + j)] = (A[(400 + j)] + A[j])
  }
}
```
i didn't change anything, please confirm if anything wrong, and what it should be

**Comment by libing4752** [comment]
I discussed with xqdan, and I think maybe we can enhance the storage_rewrite step by step. Maybe we can just update it to support case like:
`a[100]` is `float32` with length 100, `b[200]` if `float16` with length 200 , but `a` and `b` can use one buffer with size 100 * 32 bit.
so the rule is, a buffer (`A`) can merge any buffer which total size is equal to A.total_size

**Comment by xqdan** [comment]
has fixed offset when reusing different type buffer. please review @tqchen 
I'll send another PR for @libing4752 's comment, and for this case below
```
a[] = 1.0 (float32)
b[] = 1 (int16)
c[] = 1 (int 16)
d[] = b[] + c[]

b -> a[0] elem_offset = 0
c -> a[16] elem_offset = 16
```

**Comment by tqchen** [comment]
The current solution is still dangerous when dtype.bytes() and elem_type.bytes() are not dividable. Say I have a float3 and another float4. The current logic will cause problem. Also note that the real number in terms of bytes can be computed by( note that we need to multiply by lanes)

```
dtype.bits() * dtype.lanes() / 8
```

**Comment by xqdan** [comment]
Ok, for bytes calculating, i just sent  a PR to HalideIR @tqchen 
https://github.com/dmlc/HalideIR/pull/21

**Comment by tqchen** [comment]
Let us keep the byte calculation local to the functions for now, as the logic is not that hard to write

**Comment by xqdan** [comment]
i think it's better have some abstraction for some cases

**Comment by tqchen** [comment]
Both ways are fine, just since we use one way for existing cases, so we keep the convention for now

**Comment by tqchen** [comment]
I created https://github.com/dmlc/tvm/pull/885 as a basis for what we are doing in here, to simplify the code logics. elem_offset is now changed to bits_offset so the conversions are less necessary

Let us rebase against master when that is in. Now our only question would be when to reuse across buffers. In general i think @libing4752 's idea of prioritize reuse of same size is a good one.

**Comment by xqdan** [comment]
it's great, ship it please

**Pull request closed without merging**